dialogue-core/src/main/java/com/palantir/dialogue/core/CautiousIncreaseAggressiveDecreaseConcurrencyLimiter.java

```diff
@@ -122,36 +126,52 @@ void dropped() {
          */
         void success() {
             inFlight.decrementAndGet();
-            int newLimit = limit.accumulateAndGet(inFlightSnapshot, LimitUpdater.SUCCESS);
+            double newLimit = accumulateAndGetLimit(inFlightSnapshot, LimitUpdater.SUCCESS);
             if (log.isDebugEnabled()) {
                 log.debug("Increasing limit {}", SafeArg.of("newLimit", newLimit));
             }
         }
     }
 
-    enum LimitUpdater implements IntBinaryOperator {
+    enum LimitUpdater implements DoubleBinaryOperator {
         SUCCESS() {
             @Override
-            public int applyAsInt(int originalLimit, int inFlightSnapshot) {
-                if (inFlightSnapshot * 2 >= originalLimit) {
-                    return Math.min(MAX_LIMIT, originalLimit + 1);
+            public double applyAsDouble(double originalLimit, double inFlightSnapshot) {
+                if (inFlightSnapshot >= originalLimit * BACKOFF_RATIO) {
+                    // The limit is raised more easily when the maximum limit is low, and becomes linearly more
+                    // stubborn as the limit increases. Given a fixed rate of traffic this should result in
+                    // linear slope as opposed to the exponential slope expected from a static increment
+                    // value.
+                    double increment = 1D / originalLimit;
+                    return Math.min(MAX_LIMIT, originalLimit + increment);
                 }
                 return originalLimit;
             }
         },
         DROPPED() {
             @Override
-            public int applyAsInt(int originalLimit, int _inFlightSnapshot) {
-                return Math.max(MIN_LIMIT, (int) (originalLimit * BACKOFF_RATIO));
+            public double applyAsDouble(double originalLimit, double _inFlightSnapshot) {
+                return Math.max(MIN_LIMIT, originalLimit * BACKOFF_RATIO);
```

Comment:
interestingly I think taking away the cast has made this actually backoff _slower_ than before, as going 3d * 0.9 comes out as a double 2.7, whereas previously the integer cast would have taken this down to 2.0

Reply:
@carterkozak thoughts on putting a `Math.floor` on this? it would have a big effect when the limit is small but not change much when someone is going from 101 -> 90.9?

Reply:
Hah, was already running simulations on it when this came through ;-)